File: solutions/javascript/007-caesar/solution.mjs
function norm(n) {
  const m = n % 26;
  return m < 0 ? m + 26 : m;
}

function shiftChar(ch, shift) {
  const code = ch.charCodeAt(0);
  if (code >= 97 && code <= 122) return String.fromCharCode(97 + norm(code - 97 + shift));
  if (code >= 65 && code <= 90) return String.fromCharCode(65 + norm(code - 65 + shift));
  return ch;
}

export function encrypt(text, shift) {
  return [...text].map((ch) => shiftChar(ch, shift)).join("");
}

export function decrypt(text, shift) {
  return encrypt(text, -shift);
}

File: solutions/python/007-caesar/solution.py
"""Caesar cipher encrypt / decrypt."""

def _norm(n: int) -> int:
    m = n % 26
    return m + 26 if m < 0 else m


def _shift_char(ch: str, shift: int) -> str:
    if "a" <= ch <= "z":
        return chr(ord("a") + _norm(ord(ch) - ord("a") + shift))
    if "A" <= ch <= "Z":
        return chr(ord("A") + _norm(ord(ch) - ord("A") + shift))
    return ch


def encrypt(text: str, shift: int) -> str:
    return "".join(_shift_char(ch, shift) for ch in text)


def decrypt(text: str, shift: int) -> str:
    return encrypt(text, -shift)
